grails-app/services/au/org/ala/collectory/EmlImportService.groovy

```diff
@@ -137,36 +138,75 @@ class EmlImportService {
                 && eml.dataset.metadataProvider.electronicMailAddress != eml.dataset.creator.electronicMailAddress){
 
             eml.dataset.metadataProvider.each {
-                def contact = addContact(it)
+                def contact = addOrUpdateContact(it)
                 if (contact){
                     contacts << contact
                 }
             }
         }
 
-        contacts
+        // Add additional contacts
+        if (eml.dataset.contact){
+            eml.dataset.contact.each {
+                def contact = addOrUpdateContact(it)
+                if (contact){
+                    contacts << contact
+                    primaryContacts << contact
+                }
+            }
+        }
+
+        if (eml.dataset.associatedParty){
+            eml.dataset.associatedParty.each {
+                def contact = addOrUpdateContact(it)
+                if (contact){
+                    contacts << contact
+                }
+            }
+        }
+
+        [contacts: contacts, primaryContacts: primaryContacts]
     }
 
-    private def addContact(emlElement){
-        def contact = Contact.findByEmail(emlElement.electronicMailAddress)
-        if (!contact){
+    private def addOrUpdateContact(emlElement) {
+        def contact = null
+        if (emlElement.electronicMailAddress && !emlElement.electronicMailAddress.isEmpty()) {
+            String email = emlElement.electronicMailAddress.text().trim()
+            contact = Contact.findByEmail(email)
+        } else if (emlElement.individualName.givenName && emlElement.individualName.surName) {
+            contact = Contact.findByFirstNameAndLastName(emlElement.individualName.givenName, emlElement.individualName.surName)
+        } else if (emlElement.individualName.surName) {
+            // surName is mandatory
+            contact = Contact.findByLastName(emlElement.individualName.surName)
+        }
+
+        // Create the contact if it doesn't exist and it's a individualName with email or surName
+        // to prevent empty contacts (e.g. with emlElement.organizationName only)
+        boolean hasEmail = emlElement?.electronicMailAddress?.text()?.trim()?.isEmpty() == false
+        boolean hasName = emlElement?.individualName?.surName?.text()?.trim()?.isEmpty() == false
+
+        if (!contact && (hasEmail || hasName)) {
```

Comment:
Inclusion of ` && (hasEmail || hasName)` might cause null `contact` exceptions further along, if it is ever false.

Reply:
Not the case in our 500 contacts.  Looking to the eml expecification:

![image](https://github.com/AtlasOfLivingAustralia/collectory/assets/180085/b95b8847-d8e4-4004-81f7-a2ea9ff74db0)

The only case I can imagine is a contact with only organizationName and no email and no individualName that seems to me something quite useless for a "contact" (Are you suggesting that we should take into account the case of "contact: organizationName: CSIRO"?).